mock/ecs_client.go

```diff
@@ -2,67 +2,673 @@ package mock
+	if name == "" {
+		name = utility.RandomString()
+	}
+	id := arn.ARN{
+		Partition: "aws",
+		Service:   "ecs",
+		Resource:  fmt.Sprintf("container-definition:%s-%s/%s", utility.FromStringPtr(task.TaskDef.Family), name, strconv.Itoa(int(utility.FromInt64Ptr(task.TaskDef.Revision)))),
+	}
+
+	return ECSContainer{
+		ARN:      utility.ToStringPtr(id.String()),
+		TaskARN:  task.ARN,
+		Name:     def.Name,
+		Image:    def.Image,
+		CPU:      def.CPU,
+		MemoryMB: def.MemoryMB,
+	}
+}
+
+func (c *ECSContainer) export() *ecs.Container {
+	exported := &ecs.Container{
+		ContainerArn: c.ARN,
+		TaskArn:      c.TaskARN,
+		Name:         c.Name,
+		Image:        c.Image,
+	}
+
+	if c.CPU != nil {
+		exported.Cpu = utility.ToStringPtr(strconv.Itoa(int(utility.FromInt64Ptr(c.CPU))))
+	}
+	if c.MemoryMB != nil {
+		exported.Memory = utility.ToStringPtr(strconv.Itoa(int(utility.FromInt64Ptr(c.MemoryMB))))
+	}
+
+	return exported
+}
+
+func newTags(tags []*ecs.Tag) map[string]string {
+	converted := map[string]string{}
+	for _, t := range tags {
+		if t == nil {
+			continue
+		}
+		converted[utility.FromStringPtr(t.Key)] = utility.FromStringPtr(t.Value)
+	}
+	return converted
+}
+
+func exportTags(tags map[string]string) []*ecs.Tag {
+	var exported []*ecs.Tag
+	for k, v := range tags {
+		exported = append(exported, &ecs.Tag{
+			Key:   utility.ToStringPtr(k),
+			Value: utility.ToStringPtr(v),
+		})
+	}
+	return exported
+}
+
+func newEnvVars(envVars []*ecs.KeyValuePair) map[string]string {
+	converted := map[string]string{}
+	for _, envVar := range envVars {
+		if envVar == nil {
+			continue
+		}
+		converted[utility.FromStringPtr(envVar.Name)] = utility.FromStringPtr(envVar.Value)
+	}
+	return converted
+}
+
+func exportEnvVars(envVars map[string]string) []*ecs.KeyValuePair {
+	var exported []*ecs.KeyValuePair
+	for k, v := range envVars {
+		exported = append(exported, &ecs.KeyValuePair{
+			Name:  utility.ToStringPtr(k),
+			Value: utility.ToStringPtr(v),
+		})
+	}
+	return exported
+}
+
+func newSecrets(secrets []*ecs.Secret) map[string]string {
+	converted := map[string]string{}
+	for _, secret := range secrets {
+		if secret == nil {
+			continue
+		}
+		converted[utility.FromStringPtr(secret.Name)] = utility.FromStringPtr(secret.ValueFrom)
+	}
+	return converted
+}
+
+func exportSecrets(secrets map[string]string) []*ecs.Secret {
+	var exported []*ecs.Secret
+	for k, v := range secrets {
+		exported = append(exported, &ecs.Secret{
+			Name:      utility.ToStringPtr(k),
+			ValueFrom: utility.ToStringPtr(v),
+		})
+	}
+	return exported
+}
+
+// ECSService is a global implementation of ECS that provides a simplified
+// in-memory implementation of the service that only stores metadata and does
+// not orchestrate real containers or container instances. This can be used
+// indirectly with the ECSClient to access or modify ECS resources, or used
+// directly.
+type ECSService struct {
+	Clusters map[string]ECSCluster
```

Comment:
Why can't we just the `ecs` versions directly? Is it necessary to export them if we are just using the metatdata from them anyways?

Reply:
As in, use the structs from the ECS API? I thought about that, but it actually makes it more difficult to write a fake service because the payloads are not nicely formatted in a way that makes it easy to "just store the payloads and translate them into responses".

For example, the `RunTask` input includes a `TaskDefinition` string reference, which could either be the ARN or a string of "family:revision", which makes it hard to reference against an actual task definition created through `RegisterTaskDefinition`. You need to return the task definition as part of the response.